### PR TITLE
feat(stream): watermark derivation of project

### PR DESCRIPTION
**Pull request by st1page**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

implement the watermark derivation of the project.
- analyze the watermark derivation of the expression with an expression visitor.
- the frontend will send the related information in the stream project's proto
- stream project executor derives the watermark with its expression, generating a row with the input watermark

I keep it simple and stupid now. 
- the derivation only considers one expression from one input watermark column
- maybe some logic can be reused for the batch's order property derivation, but I do not consider it for now. batch's order property derivation is more complex and watermark derivation of expression is just a subset of that.
- the performance of the executor's watermark derivation is not considered an important issue. It is not the critical path and we can optimize later if we find that it is the hot spot, e.g. the simple `InputRef` expression can easily change the index and does not need to evaluate the expression.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)
close https://github.com/risingwavelabs/risingwave/issues/6813

### CODE REVIEW

**Comment by soundOfDestiny** [comment]
seems that buildkite is frozen

**Review by soundOfDestiny — APPROVED** [review]

**Review by yuhao-su — APPROVED** [review]
LGTM

File: proto/stream_plan.proto [hunk 1]
```diff
@@ -142,7 +142,13 @@ message SinkNode {
 }
 
 message ProjectNode {
+  // the three fields should have the same length
```
Comment (chenzl25) [inline proto/stream_plan.proto:145]: Three fields? Or just the last two fields should have the same length?

**Review by chenzl25 — APPROVED** [review]
LGTM!

**Review by TennyZhuang — APPROVED** [review]

**Pull request merged**